File: src/main/java/com/github/e2318501/anyban/command/BanCommand.java
package com.github.e2318501.anyban.command;

import net.md_5.bungee.api.CommandSender;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.plugin.Command;
import net.md_5.bungee.api.plugin.TabExecutor;
import com.github.e2318501.anyban.AnyBan;
import com.github.e2318501.anyban.util.Message;
import com.github.e2318501.anyban.util.TabCompleteUtil;

import java.util.*;

public class BanCommand extends Command implements TabExecutor {
    private final AnyBan plugin;

    public BanCommand(AnyBan plugin) {
        super("aban", "anyban.command.ban");
        this.plugin = plugin;
    }

    @Override
    public void execute(CommandSender sender, String[] args) {
        if (args.length >= 1) {
            String name = args[0];
            String reason = args.length == 1 ? Message.DEFAULT_BAN_REASON.text() : String.join(" ", Arrays.copyOfRange(args, 1, args.length));
            String source = sender instanceof ProxiedPlayer ? sender.getName() : Message.CONSOLE_SOURCE.text();

            Optional<UUID> uuid = plugin.getCachedPlayerManager().getUuid(name);
            if (uuid.isPresent()) {
                if (!plugin.getBannedPlayerManager().isBanned(name)) {
                    plugin.getBannedPlayerManager().add(uuid.get(), name, source, reason);

                    sender.sendMessage(Message.BAN_COMMAND.component(name, reason));

                    ProxiedPlayer proxiedPlayer = plugin.getProxy().getPlayer(name);
                    if (proxiedPlayer != null) {
                        proxiedPlayer.disconnect(Message.YOU_ARE_BANNED.component(reason));
                    }
                } else {
                    sender.sendMessage(Message.PLAYER_ALREADY_BANNED.component());
                }
            } else {
                sender.sendMessage(Message.PLAYER_NOT_FOUND.component());
            }
        } else {
            sender.sendMessage(Message.INCOMPLETE_COMMAND.component());
        }
    }

    @Override
    public Iterable<String> onTabComplete(CommandSender sender, String[] args) {
        if (args.length == 1) {
            return TabCompleteUtil.filter(plugin.getCachedPlayerManager().getNames(), args[0]);
        } else {
            return Collections.emptyList();
        }
    }
}

File: src/main/kotlin/com/github/tsuoihito/anyban/commands/BanCommand.kt
package com.github.tsuoihito.anyban.commands

import com.github.tsuoihito.anyban.AnyBan
import com.github.tsuoihito.anyban.handlers.getAddedBannedPlayerList
import com.github.tsuoihito.anyban.handlers.getNewUserUuidList
import com.github.tsuoihito.anyban.handlers.getPlayersCandidate
import com.github.tsuoihito.anyban.handlers.getUuidFromCache
import com.github.tsuoihito.anyban.utils.*
import net.md_5.bungee.api.CommandSender
import net.md_5.bungee.api.plugin.Command
import net.md_5.bungee.api.plugin.TabExecutor
import java.util.concurrent.CompletableFuture

class BanCommand(private val plugin: AnyBan) : Command("aban", "anyban.command.ban"), TabExecutor {
    override fun execute(sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            sender.sendMessage(getMessage(plugin.config, "message.invalidSyntax"))
            return
        }
        val name = args[0]
        val reason = if (args.size > 1) args[1] else plugin.config?.getString("defaultBanReason").orEmpty()

        val cachedUuid = getUuidFromCache(plugin.userUuidList, name)
        if (cachedUuid == null) {
            val fetchedUuid = fetchUuid(plugin.mapper, name, plugin.config?.getBoolean("floodgateEnabled") ?: false)
            if (fetchedUuid == null) {
                sender.sendMessage(getMessage(plugin.config, "message.invalidPlayerName") {
                    it.replace(
                        "%name%",
                        args[0]
                    )
                })
                return
            }
            plugin.userUuidList = getNewUserUuidList(
                plugin.userUuidList,
                fetchedUuid,
                name,
                calcAMonthLaterString()
            ) { calcIsExpired(it) }
            plugin.bannedPlayerList = getAddedBannedPlayerList(
                plugin.bannedPlayerList,
                fetchedUuid,
                name,
                calcCurrentTimeString(),
                sender.name,
                "forever",
                reason
            )
        }
        cachedUuid?.let {
            plugin.bannedPlayerList = getAddedBannedPlayerList(
                plugin.bannedPlayerList,
                it,
                name,
                calcCurrentTimeString(),
                sender.name,
                "forever",
                reason,
            )
        }

        sender.sendMessage(getMessage(plugin.config, "message.banSuccess") { it.replace("%name%", name) })
        kickPlayer(
            plugin.proxy,
            name,
            getMessage(plugin.config, "message.youAreBanned", false) { it.replace("%reason%", reason) })

        CompletableFuture.runAsync { plugin.saveData() }
    }

    override fun onTabComplete(sender: CommandSender, args: Array<String>): Iterable<String> {
        return if (args.size == 1) {
            getPlayersCandidate(plugin.userUuidList, args[0])
        } else listOf()
    }
}
